[WIP]Add keyboard shortcuts to file browser.

niole opened the pull request:
Resolves issue #215 

Notes to reviewers:
 * Please help suggest strategies for refactoring code to be consistent to notebook keyboard shortcuts (see `keyboardmanager.js` and `actions.js`)

minrk commented on notebook/static/tree/js/main.js:
```diff
@@ -37,6 +37,7 @@ require([
     loginwidget){
     "use strict";
 
+    //gets notebooklist creates a method
```
this comment isn't accurate

niole replied:
Thanks for the feedback. Any input on the jquery I wrote and how to
refactor it?
On Jul 21, 2015 2:34 PM, "Min RK" <notifications@github.com> wrote:

> In notebook/static/tree/js/main.js
> <https://github.com/jupyter/notebook/pull/217#discussion_r35157436>:
>
> > @@ -37,6 +37,7 @@ require([
> >      loginwidget){
> >      "use strict";
> >
> > +    //gets notebooklist creates a method
>
> this comment isn't accurate
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/notebook/pull/217/files#r35157436>.
>

minrk commented on notebook/static/base/js/namespace.js:
```diff
@@ -32,9 +32,9 @@ define(function(){
-    
+  
     //Jupyter.load_extensions = Jupyter.utils.load_extensions;
-    // 
```
Looks like you've got whitespace stripping turned on, and commits to files you didn't actually change. We would prefer to avoid committing that sort of change.

minrk commented:
We have utilities for dealing with keyboard events already that we use in the notebook, it probably doesn't make sense to write custom keyboard handling like this, with hardcoded integer keycodes, etc.

Anywhere we are going to add new keyboard shortcuts should possibly be doing it through the actions and keyboard manager stuff we set up for the notebook.

niole commented:
Thanks for the feedback. I'm going to take a look around in the code and
try to start moving things around.

On Tue, Jul 21, 2015 at 2:50 PM, Min RK <notifications@github.com> wrote:

> We have utilities for dealing with keyboard events already that we use in
> the notebook, it probably doesn't make sense to write custom keyboard
> handling like this, with hardcoded integer keycodes, etc.
>
> Anywhere we are going to add new keyboard shortcuts should possibly be
> doing it through the actions and keyboard manager stuff we set up for the
> notebook.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/notebook/pull/217#issuecomment-123487767>.
>

niole commented:
Do you think it would be better to:

 1. Create new `keyboardmanager.js`, `keyboard.js`, and `actions.js` files
for the directory tree or
 2. Implement into the existing keyboard handling utilities, using the URL
(or maybe something else e.g. event methods) to determine whether we're in
a notebook or directory tree view

I think that 1. would mean less irrelevant code gets loaded to the client
but may introduce more code duplication than 2.

On Tue, Jul 21, 2015 at 9:08 PM, Niole Nelson <niolenelson@gmail.com> wrote:

> Thanks for the feedback. I'm going to take a look around in the code and
> try to start moving things around.
>
> On Tue, Jul 21, 2015 at 2:50 PM, Min RK <notifications@github.com> wrote:
>
>> We have utilities for dealing with keyboard events already that we use in
>> the notebook, it probably doesn't make sense to write custom keyboard
>> handling like this, with hardcoded integer keycodes, etc.
>>
>> Anywhere we are going to add new keyboard shortcuts should possibly be
>> doing it through the actions and keyboard manager stuff we set up for the
>> notebook.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/jupyter/notebook/pull/217#issuecomment-123487767>.
>>
>
>

niole commented:
Do you think it would be better to:

 1. Create new `keyboardmanager.js`, `keyboard.js`, and `actions.js` files
for the directory tree or
 2. Implement into the existing keyboard handling utilities, using the URL
(or maybe something else e.g. event methods) to determine whether we're in
a notebook or directory tree view

I think that 1. would mean less irrelevant code gets loaded to the client
but may introduce more code duplication than 2.

niole commented:
resolves issue #215

feynmanliang commented on notebook/Untitled.ipynb:
```diff
@@ -0,0 +1,6 @@
+{
```
Delete this file

feynmanliang commented on notebook/Untitled1.ipynb:
```diff
@@ -0,0 +1,65 @@
+{
```
Delete this file

feynmanliang commented on notebook/base/Untitled.ipynb:
```diff
@@ -0,0 +1,6 @@
+{
```
Delete this file

feynmanliang commented on notebook/static/notebook/js/keyboardmanager.js:
```diff
@@ -25,24 +25,32 @@ define([
+          this.tree_shortcuts = new keyboard.ShortcutManager(undefined, options.events, this.actions, this.env);
+          this.tree_shortcuts.add_shortcuts(this.get_default_tree_shortcuts());
+        }
+        if (this.view === 'notebook'){
```
`else if`

feynmanliang replied:
Actually since you set `this.view = 'tree'` on L32, this branch will never execute. Can you test to make sure that notebook view shortcuts have not regressed?

niole replied:
the way things update, this.view updates the way you want it to. you won't
run into problems

On Tue, Aug 4, 2015 at 5:48 PM, Feynman Liang <notifications@github.com>
wrote:

> In notebook/static/notebook/js/keyboardmanager.js
> <https://github.com/jupyter/notebook/pull/217#discussion_r36259861>:
>
> >          this.last_mode = undefined;
> >          this.bind_events();
> >          this.env = {pager:this.pager};
> >          this.actions = options.actions;
> > -        this.command_shortcuts = new keyboard.ShortcutManager(undefined, options.events, this.actions, this.env );
> > -        this.command_shortcuts.add_shortcuts(this.get_default_common_shortcuts());
> > -        this.command_shortcuts.add_shortcuts(this.get_default_command_shortcuts());
> > -        this.edit_shortcuts = new keyboard.ShortcutManager(undefined, options.events, this.actions, this.env);
> > -        this.edit_shortcuts.add_shortcuts(this.get_default_common_shortcuts());
> > -        this.edit_shortcuts.add_shortcuts(this.get_default_edit_shortcuts());
> > +        if (this.view === 'tree'){
> > +          this.tree_shortcuts = new keyboard.ShortcutManager(undefined, options.events, this.actions, this.env);
> > +          this.tree_shortcuts.add_shortcuts(this.get_default_tree_shortcuts());
> > +        }
> > +        if (this.view === 'notebook'){
>
> else if
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/notebook/pull/217/files#r36259861>.
>

feynmanliang replied:
I see now, it's set inside `bind_events()`. I think that L32 belongs inside `bind_events()` to keep logic centralized

feynmanliang commented on notebook/static/notebook/js/keyboardmanager.js:
```diff
@@ -60,6 +68,16 @@ define([
+
+    KeyboardManager.prototype.get_default_tree_shortcuts = function() {
+        return {
+            'enter'       : 'ipython.enter-row',
```
Align on `:`

feynmanliang commented on notebook/static/notebook/js/keyboardmanager.js:
```diff
@@ -25,24 +25,32 @@ define([
          *    @param options.pager: {Pager}  pager instance
          */
         this.mode = 'command';
+        this.view = 'tree';
```
Move this into `KeyboardManager.prototype.bind_events()`

takluyver replied:
Actually, I'd rather rethink this to produce a cleaner architecture.

There should be a lower-level keyboard piece of code that handles the basic architecture of listening for shortcuts and responding to them, and then the notebook and the tree view should be able to build on that separately to listen for the shortcuts they want. You may have to split up and refactor some of the code to do this.

However, there's some big refactorings going on for 5.0, and I don't know if this code will still be used in its present form. @Carreau , @jdfreder : do either of you know what the fate of the keyboard interaction code will be?

jdfreder replied:
The keyboard interaction will change I think, but in time.  I think while we are still maintaining this branch, we might as well consider allowing this feature to be merged.  I'm of a similar mindset to @Carreau and am all for the addition of his and @svurens command palette, despite the refactor that is happening.

feynmanliang commented on notebook/static/notebook/js/keyboardmanager.js:
```diff
@@ -123,7 +141,14 @@ define([
 
     KeyboardManager.prototype.bind_events = function () {
         var that = this;
+        var notebook_view = window.document.getElementById("pager");
```
This doesn't seem to be a very robust way to test whether we are in notebook or tree view (a change to the HTML will break this logic). Is there a better way to test for this?

If not, please add a comment that this test expects "pager" to only exists when in notebook view.

feynmanliang commented on notebook/static/notebook/js/keyboardmanager.js:
```diff
@@ -145,25 +175,27 @@ define([
+
+        if (this.view === 'tree'){
+          this.tree_shortcuts.call_handler(event);
+        }else{
```
space between `}` and `else`

feynmanliang commented on notebook/static/notebook/js/keyboardmanager.js:
```diff
@@ -145,25 +175,27 @@ define([
+              }
+              return true;
+          }
+          if (this.mode === 'edit') {
```
This logic is only present in notebook view (tree view has no modes); move into the else branch on the `this.view === 'tree'` conditional

feynmanliang commented on notebook/static/tree/js/main.js:
```diff
@@ -54,18 +60,32 @@ require([
         common_config: common_config
     });
+
+    page = new page.Page();
```
Move back to L42; in general, minimize unnecessary changes to codebase

feynmanliang commented on notebook/static/tree/js/notebooklist.js:
```diff
@@ -315,6 +316,35 @@ define([
+    NotebookList.prototype.is_selected = function(){
+      if ($('.focus').size() === 0){
+        $('.list_item').first().attr("tabindex","0").addClass('focus').focus();
+      }else{
```
space before `else`

feynmanliang commented on notebook/static/tree/js/notebooklist.js:
```diff
@@ -315,6 +316,35 @@ define([
         this._selection_changed();  
     };
 
+    NotebookList.prototype.is_selected = function(){
```
This isn't really `is_selected` since there is a side effect to focus on the first list item if nothing is in focus. Where is this function used?

feynmanliang commented on notebook/static/tree/js/treeactions.js:
```diff
@@ -0,0 +1,322 @@
+        }
+    };
+
+    /**
```
Are any of these actions used in NotebookList? If not, remove from this file.

feynmanliang commented on notebook/templates/notebook.html:
```diff
@@ -324,6 +324,6 @@
 <script src="{{ static_url("components/text-encoding/lib/encoding.js") }}" charset="utf-8"></script>
 
-<script src="{{ static_url("notebook/js/main.min.js") }}" charset="utf-8"></script>
+<script src="{{ static_url("notebook/js/main.js") }}" charset="utf-8"></script>
```
Point towards minified version

feynmanliang commented:
Please include a summary of the changes (e.g. added `tree_shortcuts` to `KeyboardManager`, copied `actions.js` -> `treeactions.js` and made modifications XYZ, etc). Also state the default bindings provided in `get_default_tree_shortcuts`.

niole commented:
@minrk @takluyver please review.
Overview of changes/additions:

  In `tree/js/main.js`, initialize new `treeactions.js` file via `actions.init()`, pass into a `KeyboardManager` instance, which is then passed into a `NotebookList` instance. The new `notebook_list` object is then added to the `env` local to `ActionHandler` in `treeaction.js`, via a `.set_notebooklist` function I added to the `KeyboardManager` prototype.
  Added new functions to the `NotebookList` prototype to handle keyboard events:
`.get_selected_index`
`.get_length_siblings`
`.select_row`
`.open`
  Created  the actions file for instances of notebooklist, `tree/js/treeactions.js`, which is a copy of `notebook/js/actions.js`, but has different enters in the `_actions` object.
  Created a set of default shortcuts for the file tree by adding `.get_default_tree_shortcuts` to the `KeyboardManager` prototype, which returns action names for 'enter', 'j', and 'k'.
  Also, added a `view` variable to the `KeyboardManager` prototype in order to differentiate when to build shortcuts for the file tree or for a newly created notebook. This variable is updated in `KeyboardManager.prototype.bind_events`.

takluyver commented on notebook/static/tree/js/treeactions.js:
```diff
@@ -0,0 +1,322 @@
+     **/
+    var _actions = {
+        'enter-row': {
+            help    : 'enters row',
```
I'm not sure what this means. 'Opens current file'?

takluyver commented on notebook/static/tree/js/treeactions.js:
```diff
@@ -0,0 +1,322 @@
+            }
+        },
+        'select-previous-row':{
+            help    : 'selects previous row',
```
On the notebook list, 'select' to me means checking the checkbox next to an item. I'd describe what this does in terms of a 'cursor' or 'focus', e.g. 'move cursor down'.

takluyver commented on notebook/templates/tree.html:
```diff
@@ -167,5 +167,5 @@
     {{super()}}
 
-<script src="{{ static_url("tree/js/main.min.js") }}" type="text/javascript" charset="utf-8"></script>
+<script src="{{ static_url("tree/js/main.js") }}" type="text/javascript" charset="utf-8"></script>
```
It's fine to change this for testing, but please don't commit the change.

takluyver commented:
As @minrk and @feynmanliang have pointed out, there's quite a few unwanted changes that have got into this PR - files added and whitespace-only changes. We'll need to rebase this to trim those changes out.

niole commented:
* Thanks for the requested changes. I'm going to start working on it for
the sake of cleaning up the code, but I'm curious to hear about the
refactorings for the 5.0 milestone/if this PR bears significance therein.
* the `'enter-row'` action name in `_actions`  was made to open the focused
file/directory on enter's keydown
* I'm also curious to know if anyone has an opinion on `this.view` in
`KeyboardManager`, as in should it be initially defined in the
`KeyboardManager` constructor or in
`KeyboardManager.prototype.bind_events`, which is where it gets assigned.
*  Also wondering if anyone has an opinion on the location of the `if
(this.view === 'tree')` statement on line 186 in `keyboardmanager.js`,

On Wed, Aug 5, 2015 at 11:37 AM, Thomas Kluyver <notifications@github.com>
wrote:

> As @minrk <https://github.com/minrk> and @feynmanliang
> <https://github.com/feynmanliang> have pointed out, there's quite a few
> unwanted changes that have got into this PR - files added and
> whitespace-only changes. We'll need to rebase this to trim those changes
> out.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/notebook/pull/217#issuecomment-128102968>.
>

takluyver commented:
See my inline comment - I think it would be best to refactor things a bit so the keyboard manager doesn't have a `this.view` or conditional logic based on it.

niole commented:
Gotcha
On Aug 5, 2015 12:24 PM, "Thomas Kluyver" <notifications@github.com> wrote:

> See my inline comment - I think it would be best to refactor things a bit
> so the keyboard manager doesn't have a this.view or conditional logic
> based on it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/notebook/pull/217#issuecomment-128118437>.
>

niole commented:
@takluyver @mink reset unnecessary commits and, renamed action names in treeactions.js to be  more descriptive, removed unused action handlers in notebooklist.js, moved this.view into KeyboardManager.bind_events to help clarify it's use case

niole commented:
@takluyver @mink Should I close this?

rgbkrk commented:
Min and Thomas have been transitioning to the other side of the globe, I'd just wait a bit for them to get back.

minrk commented:
@niole there's still quite a bit of work to do on this PR to get it into working shape, including addressing the comments about avoiding committing the extra unintended changes. If you aren't prepared to finish the work, we can close it here, for now. You are always free to try again another time, but I would recommend starting with something smaller. Keyboard shortcuts in browsers are one of the more complicated and hairy aspects of the notebook.

niole commented:
@mink I rebased and squashed the commit history. let me know how it's looking.

minrk commented:
@niole sorry, I don't think this has been rebased correctly. It now has many commits from others, going back months. Perhaps you rebased it on the wrong branch? The commit it should be based on (as of right now) is d43d72f10d8a425432351222129d9100e64e8827.

niole commented:
@minrk you're right, I should've checked twice. I think it's fixed now!

Carreau commented:
This would need a rebase (and removing WIP from title). 
It seem it has been forgotten as it slipped on the second page of PR list. Sorry about that.

niole commented:
@Carreau , I recall that when I put this up, Jupyter was considering a widespread refactor of the entire app. I can't imagine a rebase would bring this into a mergeable staging. Closing for now.

Carreau commented:
> @Carreau , I recall that when I put this up, Jupyter was considering a widespread refactor of the entire app. I can't imagine a rebase would bring this into a mergeable state. Closing for now.

ok, I'll try to have a look at how hard a rebase would be !
Sorry for that.
